indra/llcommon/llcallbacklist.cpp

```diff
@@ -45,186 +44,343 @@ LLCallbackList::~LLCallbackList()
+    // call in tick()? How to map from the top() value back to the
+    // corresponding handle_t? That's why we store func_at::mToken.
+
+    // fibonacci_heap provides a pair of begin()/end() methods to iterate over
+    // all nodes (NOT in heap order), plus a function to convert from such
+    // iterators to handles. Without mMeta, that would be our only chance
+    // to validate.
+    auto found{ mMeta.find(timer.token) };
+    if (found == mMeta.end())
+    {
+        // we don't recognize this handle -- maybe the timer has already
+        // fired, maybe it was previously canceled.
+        return false;
+    }
+
+    // Funny case: what if the callback directly or indirectly reaches a
+    // cancel() call for its own handle?
+    if (found->second.mRunning)
+    {
+        // tick() has special logic to defer the actual deletion until the
+        // callback has returned
+        found->second.mCancel = true;
+        // this handle does in fact reference a live timer,
+        // which we're going to cancel when we get a chance
+        return true;
+    }
+
+    // Erase from mQueue the handle_type referenced by timer.token.
+    mQueue.erase(found->second.mHandle);
+    // before erasing the mMeta entry
+    mMeta.erase(found);
+    if (mQueue.empty())
+    {
+        // If that was the last active timer, unregister for callbacks.
+        //LLCallbackList::instance().deleteFunction(mLive);
+        // Since we're in the source file that knows the true identity of an
+        // LLCallbackList::handle_t, we don't even need to call instance().
+        mLive.disconnect();
+    }
+    return true;
+}
+
+// RAII class to set specified variable to specified value
+// only for the duration of containing scope
+template <typename VAR, typename VALUE>
+class TempSet
 {
 public:
-	BoolFuncEventTimer(bool_func_t callable, F32 seconds):
-		LLEventTimer(seconds),
-		mCallable(callable)
-	{
-	}
-private:
-	BOOL tick()
-	{
-		return mCallable();
-	}
+    TempSet(VAR& var, const VALUE& value):
+        mVar(var),
+        mOldValue(mVar)
+    {
+        mVar = value;
+    }
+
+    TempSet(const TempSet&) = delete;
+    TempSet& operator=(const TempSet&) = delete;
 
-	bool_func_t mCallable;
+    ~TempSet()
+    {
+        mVar = mOldValue;
+    }
+
+private:
+    VAR& mVar;
+    VALUE mOldValue;
 };
 
-// Call a given callable every specified number of seconds, until it returns true.
-void doPeriodically(bool_func_t callable, F32 seconds)
+bool Timers::tick()
 {
-	new BoolFuncEventTimer(callable, seconds);
+    // Fetch current time only on entry, even though running some mQueue task
+    // may take long enough that the next one after would become ready. We're
+    // sharing this thread with everything else, and there's a risk we might
+    // starve it if we have a sequence of tasks that take nontrivial time.
+    auto now{ LLDate::now().secondsSinceEpoch() };
+    auto cutoff{ now + TIMESLICE };
```

Comment:
I suggest exposing this to callbacks/Timers in some fassion. A lot of the 'onIdle' callbacks are declaring own slices. Ex: inventory search and population, which means every visible inventory panel uses own 'slice'.

Instead of making TIMESLICE constant I suggest making it dynamic. Ex: when loading bar is visible (during teleport), may be when viewer is in the background allow larger slices since fps won't matter as much.

Reply:
Interesting point. Okay. I can support runtime get/set operations, but I don't propose to make those calls in the example places you mentioned.

I'm glad you're not suggesting that LLCallbackList try to vary the timeslice depending on the nature of the tasks in the queue.

Reply:
Please see [this commit](https://github.com/secondlife/viewer/pull/1555/commits/ce9886e3d374df3017eeaeeb9e30f28343712896) on the lua-timers branch.